controllers/bpfman-agent/application-program_test.go

```diff
@@ -0,0 +1,308 @@
+package bpfmanagent
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	bpfmaniov1alpha1 "github.com/bpfman/bpfman-operator/apis/v1alpha1"
+	bpfmanagentinternal "github.com/bpfman/bpfman-operator/controllers/bpfman-agent/internal"
+	agenttestutils "github.com/bpfman/bpfman-operator/controllers/bpfman-agent/internal/test-utils"
+	"github.com/bpfman/bpfman-operator/internal"
+	testutils "github.com/bpfman/bpfman-operator/internal/test-utils"
+	gobpfman "github.com/bpfman/bpfman/clients/gobpfman/v1"
+	"github.com/google/go-cmp/cmp"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/protobuf/testing/protocmp"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes/scheme"
+)
+
+func TestBpfApplicationControllerCreate(t *testing.T) {
+	var (
+		// global config
+		name         = "fakeAppProgram"
+		namespace    = "bpfman"
+		bytecodePath = "/tmp/hello.o"
+		fakeNode     = testutils.NewNode("fake-control-plane")
+		ctx          = context.TODO()
+		// fentry program config
+		bpfFentryFunctionName = "fentry_test"
+		fentryFunctionName    = "do_unlinkat"
+		fentryBpfProgName     = fmt.Sprintf("%s-%s-%s-%s", name, "fentry", fakeNode.Name, "do-unlinkat")
+		fentryBpfProg         = &bpfmaniov1alpha1.BpfProgram{}
+		fentryFakeUID         = "ef71d42c-aa21-48e8-a697-82391d801a81"
+		// kprobe program config
+		bpfKprobeFunctionName       = "kprobe_test"
+		kprobeFunctionName          = "try_to_wake_up"
+		kprobeBpfProgName           = fmt.Sprintf("%s-%s-%s-%s", name, "kprobe", fakeNode.Name, "try-to-wake-up")
+		kprobeBpfProg               = &bpfmaniov1alpha1.BpfProgram{}
+		kprobeFakeUID               = "ef71d42c-aa21-48e8-a697-82391d801a82"
+		kprobeOffset                = 0
+		kprobeRetprobe              = false
+		kprobecontainerpid    int32 = 0
+	)
+
+	// A AppProgram object with metadata and spec.
+	App := &bpfmaniov1alpha1.BpfApplication{
```

Comment:
I think we should test at least 3 program types here.  I'd actually like to test all of them becasue they all have unique code that is almost the same and mistakes can creep in if some are updated but not all.  However, these unit tests are laborious, so it would be okay to test all types in an integration test.